crypto/src/mls/session/key_package.rs

```diff
@@ -248,25 +250,32 @@ impl Session {
         &self,
         kps: &'a [(MlsKeyPackage, KeyPackage)],
         keystore: &CryptoKeystore,
-        refs: &[KeyPackageRef],
-    ) -> Result<Vec<&'a [u8]>, Error> {
-        let kp_to_delete: Vec<_> = kps
-            .iter()
-            .filter_map(|(store_kp, kp)| {
-                let is_expired = Self::is_mls_keypackage_expired(kp);
-                let mut to_delete = is_expired;
-                if !(is_expired || refs.is_empty()) {
-                    // not expired and there are some refs to check
-                    // then delete it when it's found in the refs
-                    to_delete = refs.iter().any(|r| r.as_slice() == store_kp.keypackage_ref);
-                }
-
-                to_delete.then_some((kp, &store_kp.keypackage_ref))
+        refs: impl IntoIterator<Item = KeyPackageRef>,
+    ) -> Result<HashSet<&'a [u8]>, Error> {
+        let refs = refs
+            .into_iter()
+            .map(|kp| {
+                // If `KeyPackageRef` implemented `Hash + PartialEq<Rhs=[u8]> + Eq`, then we could just check whether
+                // an arbitrary reference existed in the hashset without moving data here at all; the type could just
+                // be `HashSet<KeyPackageRef>`.
+                //
+                // If `KeyPackageRef` implemented `fn into_inner(self) -> Vec<u8>` then we could at least extract the
+                // data without copying.
+                //
+                // As things stand, we're stuck with some pointless copying of (usually short) data around.
+                // Hopefully LLVM is smart enough to optimize some of it away!
+                kp.as_slice().to_owned()
             })
-            .collect();
+            .collect::<HashSet<_>>();
+
+        let kp_to_delete = kps.iter().filter_map(|(store_kp, kp)| {
+            let is_expired = Self::is_mls_keypackage_expired(kp);
+            let to_delete = is_expired || refs.contains(store_kp.keypackage_ref.as_slice());
+            to_delete.then_some((kp, &store_kp.keypackage_ref))
+        });
 
-        for (kp, kp_ref) in &kp_to_delete {
-            // TODO: maybe rewrite this to optimize it. But honestly it's called so rarely and on a so tiny amount of data. Tacking issue: WPB-9600
+        // note: we're cloning the iterator here, not the data
+        for (kp, kp_ref) in kp_to_delete.clone() {
```

Comment:
Here we're removing a reference to WPB-9600, but that item is still open. We should probably close it.